File: runtime/peerbot/src/newbound/p2p/protocol/readkey.py
from newbound.crypto.supersimplecipher import SuperSimpleCipher

class READKEY(object):
	def __init__(self, service):
		self.service = service
		
	def execute(self, data):
		uuid = data['uuid']
		ba = data['data']
		sock = data['p2psocket']
		p = self.service.p2p.getPeer(uuid)
		if p.readkey == None:
			print("Got read key for "+uuid)
			ba = self.service.p2p.decryptWithPrivateKey(uuid, ba)
			p.readkey = SuperSimpleCipher(ba)
			if p.port == -1 or p.port == 0: p.port = sock.parser.remoteport
			self.service.p2p.peers.savePeer(p)
		else:
			print("Got read key for "+uuid+", but we already have one. Ignoring.")
		sock.connecting = False
		self.service.p2p.isOK(uuid, sock, None, None)

File: runtime/peerbot/src/com/newbound/p2p/protocol/READKEY.java
package com.newbound.p2p.protocol;

import org.json.JSONObject;

import com.newbound.p2p.P2PPeer;
import com.newbound.p2p.P2PService;
import com.newbound.p2p.P2PSocket;
import com.newbound.robot.Callback;

public class READKEY implements Callback {

	P2PService P2P;

	public READKEY(P2PService p2p) 
	{
		super();
		P2P = p2p;
	}

	@Override
	public void execute(JSONObject data) 
	{
		try
		{
			String uuid = data.getString("uuid");
			byte[] ba = (byte[])data.get("data");
			P2PSocket sock = (P2PSocket)data.get("p2psocket");
			P2PPeer p = P2P.getPeer(uuid);
			
			System.out.println("Got read key "+uuid);
			ba = P2P.decryptWithPrivateKey(uuid, ba, 0, ba.length);
			p.setReadKey(ba);
			sock.CONNECTING = false;
			P2P.isOK(uuid, sock, null, null);
		}
		catch (Exception x) { x.printStackTrace(); }
	}

}
